### ISSUE TITLE
BsForm "Cannot read property 'dispatchEvent' of undefined"

### ISSUE THREAD

**Issue by basz** [kaliber5/ember-bootstrap#1181]
I'm seeing `Cannot read property 'dispatchEvent' of undefined` for forms with submitOnEnter. 

It comes from here;

```
@action
  handleKeyPress(e) {
    let code = e.keyCode || e.which;
    if (code === 13 && this.submitOnEnter) {
      this.triggerSubmit();
    }
  }

  triggerSubmit() {
    let event = document.createEvent('Event');
    event.initEvent('submit', true, true);
    this._element.dispatchEvent(event);
  }
```

this._element is null, hence the error.
_element is set in the form template with the ref helper...

```
  {{ref this "_element"}}
  {{on "keypress" this.handleKeyPress}}
  {{on "submit" this.handleSubmit}}
```

Since the error happening depends on the path the user takes through the application, sometimes _element is correct. It is consistently erroring for those paths.

I've fiddled a bit with using the `fn` helper. Also, perhaps we can simply use `event.target` as that is suppost to be the element... Dunno, curious to see if anyone else has this problem...

---
Output from `ember bootstrap:info`:

```
Npm packages:
ember-bootstrap: ^4.0.0 -> 4.0.0
ember-cli: ~3.20 -> 3.20.0
bootstrap: ^4.4.1 -> 4.5.0
bootstrap-sass: n/a
ember-cli-sass: ^9.0.0 -> 10.0.1
ember-cli-less: n/a
Bower packages:
bootstrap: n/a
bootstrap-sass: n/a
ember-bootstrap configuration:
bootstrapVersion: 4
importBootstrapCSS: false
importBootstrapTheme: false
importBootstrapFont: false
whitelist: undefined
```

Edit: I do not have `ember-ref-bucket` installed as was suggested.

**Comment by jelhan** [comment]
Is #1182 fixing the issue for you?

**Comment by jelhan** [comment]
Another solution may be to inline `triggerSubmit` in `handleKeyPress`. It doesn't seem to be used anywhere else. And we could dispatch the event on `event.target` of the key press event. No need to use `this._element` at all. Would also simplify implementation, wouldn't it?

**Comment by basz** [comment]
Yes, I believe so...

**Comment by basz** [comment]
> Is #1182 fixing the issue for you?

It won't run here... possibly https://github.com/lifeart/ember-ref-bucket/issues/1

**Comment by basz** [comment]
I've now made a change to bs-form (locally) to remove the ref helper for this case as suggested. Seems to work, but am not sure if the signature change might be a BC. @jelhan thoughts?

```
  @action
  handleKeyPress(event) {
    let code = event.keyCode || event.which;
    if (code === 13 && this.submitOnEnter) {
      this.triggerSubmit(event.target);
    }
  }

  triggerSubmit(target) {
    let event = document.createEvent('Event');
    event.initEvent('submit', true, true);
    target.dispatchEvent(event);
  }
```

also: the ref helper is used in more places that might need to be refactored then... https://github.com/kaliber5/ember-bootstrap/search?q=%22%7B%7Bref%22&unscoped_q=%22%7B%7Bref%22&type=Code

maybe wait until the ref helper is released?

**Comment by jelhan** [comment]
Yes. Actually I would even changed it to:

```js
  @action
  handleKeyPress(event) {
    let code = event.keyCode || event.which;
    if (code === 13 && this.submitOnEnter) {
      let submitEvent = document.createEvent('Event');
      submitEvent.initEvent('submit', true, true);
      event.target.dispatchEvent(submitEvent);
    }
  }
```

Less code.

**Comment by simonihmig** [comment]
Closed by #1183
